FIREFLY-2033: React Component Library for Firefly Visualizations (#1970)
Caltech-IPAC/firefly: component-library/README.md

# Firefly Component Library

A React component library for astronomical data visualization, built on
[Firefly](https://github.com/Caltech-IPAC/firefly), the web visualization

> framework from [IPAC / Caltech](https://www.ipac.caltech.edu/).

Reviewer: ```suggestion
framework from [Caltech/IPAC](https://www.ipac.caltech.edu/).
```

that's what the recommended way is for written materials.

The library provides ready-to-use components for displaying catalog tables,

> interactive charts, and FITS sky images that integrate with a running Firefly

Reviewer: Maybe mention: FITS and HiPS sky images? or astronomical images?

server.

---

## Installation

```bash
yarn add @ipac/firefly-component-library
```

### Peer dependencies

> The library requires these packages in your project:

Reviewer: "requires" sound like direct dependencies - maybe say something like the following (drafted with AI):

> These packages are expected to be provided by the host application, allowing our library to share your project's existing instance of React, Joy UI, etc. without duplicating core code or breaking your global theme context:

ReactDOM.createRoot(document.getElementById('root')).render(
  <CssVarsProvider theme={theme}>
    <App />
  </CssVarsProvider>
);
```

---

## Available components

The library is split into subpath packages so you only import what you need.

**`@ipac/firefly-component-library`** is the core package. It contains `initFirefly`,
the Firefly theme, form inputs, and field group utilities.

**`@ipac/firefly-component-library/tables`** contains catalog table components and
utilities for loading, searching, and manipulating tabular data.

**`@ipac/firefly-component-library/charts`** contains interactive chart components
(scatter plots, histograms, and more) that link directly to table data.

**`@ipac/firefly-component-library/images`** contains FITS image viewer components
for sky image display with WCS overlays and color stretch controls.

```js
import { initFirefly }       from '@ipac/firefly-component-library';
import { DataTable }         from '@ipac/firefly-component-library/tables';
import { HistogramChart }    from '@ipac/firefly-component-library/charts';
import { ImagePlot }         from '@ipac/firefly-component-library/images';
```

---

## Try it locally

> This walks through creating a minimal React app that renders three linked components: a FITS image viewer on the left, a scatter chart on the right, and a catalog table at the bottom, all driven by the same WISE dataset.

Reviewer: ```suggestion
This walks through creating a minimal React app that renders three linked components: a survey map (HiPS) viewer on the left, a scatter chart on the right, and a catalog table at the bottom, all driven by the same WISE dataset.
```

since "coverage" viewer isn't FITS image.

### 1. Create the app

```bash
yarn create vite test-firefly-comp-lib --template react
```

Do not run `yarn dev` yet — the library and its peer dependencies still need to be installed.

### 2. Install peer dependencies

```bash
cd test-firefly-comp-lib
```

See [Peer dependencies](#peer-dependencies) above for the full install command.

### 3. Install the library

If you have a local build:

```bash
# from component-library/
yarn build && yarn pack
# produces package.tgz

cd /path/to/test-firefly-comp-lib
yarn add /path/to/component-library/package.tgz
```

Or once published:

```bash
yarn add @ipac/firefly-component-library

> ```

Reviewer: We should move this to the top of this section to avoid confusion for the end users. And then the local build instructions can become a subsection like "development install"

    <CssVarsProvider theme={extendTheme(defaultTheme())}>
        <App />
    </CssVarsProvider>
);
```

### 5. Create the welcome page

Create `src/App.jsx`:

```jsx
import React from 'react';
import {DataTable, makeFileRequest}
    from '@ipac/firefly-component-library/tables';
import {ScatterChart}
    from '@ipac/firefly-component-library/charts';
import {Coverage}
    from '@ipac/firefly-component-library/images';

const TABLE_URL = 'https://web.ipac.caltech.edu/staff/roby/demo/WiseDemoTable.tbl';
const TBL_ID = 'wise-demo';

const WISE_REQUEST = makeFileRequest('WISE Table', TABLE_URL, null, {tbl_id: TBL_ID});

export function App() {
    return (
        <div style={{display: 'flex', flexDirection: 'column', height: '100vh', padding: 8, gap: 8}}>
            <h2 style={{margin: 0}}>Firefly Component Library Demo</h2>
            <div style={{display: 'flex', gap: 8, flex: 1, minHeight: 0}}>
                <Coverage sx={{flex: 1}}/>
                <ScatterChart tbl_id={TBL_ID} x_axis='crval1' y_axis='crval2' sx={{flex: 1}}/>
            </div>
            <DataTable source={WISE_REQUEST} sx={{flex: '0 0 280px'}}/>
        </div>
    );
}
```

> Selecting or filtering rows in the table highlights the corresponding points in the scatter chart because they share the same `tbl_id`.

Reviewer: ```suggestion
Selecting or filtering rows in the table highlights the corresponding points in the scatter chart and corresponding overlay markers in the coverage image because they share the same `tbl_id`.
```
